pkg/sql/sem/tree/type_check.go

```diff
@@ -546,6 +546,25 @@ func (expr *CollateExpr) TypeCheck(ctx *SemaContext, desired *types.T) (TypedExp
 		"incompatible type for COLLATE: %s", t)
 }
 
+// AdjustTypeForCollate returns the collated version of the input type.
+func AdjustTypeForCollate(typ *types.T, locale string) (*types.T, error) {
+	switch typ.Family() {
+	case types.StringFamily:
+		return types.MakeCollatedString(typ, locale), nil
+	case types.CollatedStringFamily:
+		return nil, pgerror.Newf(pgcode.Syntax, "multiple COLLATE declarations")
+	case types.ArrayFamily:
+		elemTyp, err := AdjustTypeForCollate(typ.ArrayContents(), locale)
```

Comment:
can you double check if this works for 2D arrays?
(do we even support 2D arrays?)

Reply:
We don't support multidimensional arrays but I think this still works in that case (I can double check).